[cmake] Modify file output and opencv download method

zmoth opened the pull request:
* Modify the download method of opencv in cmake
Use the variable `USE_LOCAL_OPENCV_PACKAGE` to decide whether to download opencv, and add a sha256 checksum
* Modify output path
Has consistent output on both windows and ubuntu.
```
├─bin
│  ├─Debug
│  │  └─Plugins
│  ├─Release
│  │  └─Plugins
│  └─RelWithDebInfo
│      └─Plugins
├─lib
│  ├─Debug
│  │  └─Plugins
│  ├─Release
│  │  └─Plugins
│  └─RelWithDebInfo
│      └─Plugins
```
* Set cmake minimum version to 3.16
No problem on windows and ubuntu

* Compile with Release, remove the console window, in the windows environment

Sorry, I don't have a maxos environment.

It may be a big change, but I've separated them into different commits @rwardlow01

zmoth committed:
Modify the download method of opencv in cmake

Turn on all build options by default
Adding a suffix to the generated library

zmoth committed:
Set cmake minimum version to 3.16

zmoth committed:
Modify output path

zmoth committed:
When compiling with Release, the console window is not generated

rwardlow01 commented:
Some comments I have for this PR:

I think USE_LOCAL_OPENCV_PACKAGE should be OFF by default because with it ON by default it won't automatically download the necessary package and I would have to know or check the cmake file to know what's going on and provide the proper opencv path, I usually like the default behavior to just work and allow the developer to add overrides if they want once they are familiar with the build process. I think having that in the first place is kind of redundant though which is why I didn't create a variable like that. You can already achieve the same effect by simply setting cmake option -DOpenCV_DIR=/your/opencv/path and if the cmake script sees that set and finds OpenCV on the initial check then it uses your specified version otherwise it will download the one online and extract it (the proper default behavior).

It looks like you stripped out all of the RPath settings, the reason those are in there is for package building convenience, this allows the executable to properly find the dynamic libraries within the app folder and for plugins it allows the plugin to find any of its dependency libraries within the app or its own plugin folder.
see: https://cmake.org/cmake/help/latest/prop_tgt/INSTALL_RPATH.html
and for MacOS it uses INSTALL_NAME_DIR, see: https://cmake.org/cmake/help/latest/prop_tgt/INSTALL_NAME_DIR.html

On windows I had it copying the OpenCV DLL files into the executable build dir so you could easily run it without having to setup the path in the environment variable to find the DLLs on the search path but it seems you removed that so it no longer copies the DLLs, why is that? this could perhaps be a case for a cmake variable OPENCV_COPY_DLLS or something like that.

On Linux it is not building into the bin/Release and bin/Release/Plugins folders for me for some reason, everything just ends up in the root of the bin folder.
I also think creating a Release dir inside bin is redundant because you already have your separate cmake folders for release and debug for example cmake-build-release and cmake-build-debug or whatever you want to name them so you would end up with a release/bin/release type of thing which doesn't make sense.

I don't mind the console window coming up on the release build at least until I get a proper logging system implemented because right now it's the only way to convey messages to the user in case of error or other feedback in this early stage of development.

I haven't had a chance to test your changes on MacOS yet, I assume there will be similar issues with the INSTALL_NAME_DIR path being removed and possibly the build path changes, I'll let you know what I find as soon as I have some more time.

zmoth commented:
Hello @rwardlow01 ，Because I am using a translation software, I may not understand it and speak imprudent. :)

The editor I use is vscode.

## USE_LOCAL_OPENCV_PACKAGE
I don't really like the `-DOpenCV_DIR=/opencv/path` way to build, I prefer to do it in cmake. Because I have to configure it again when I clear the cache.
Now I've modified part of the logic.

- If `USE_LOCAL_OPENCV_PACKAGE=OFF`, then opencv will **NOT** be checked locally and will be downloaded and added as default.
- If `USE_LOCAL_OPENCV_PACKAGE=ON`, first check whether opencv exists locally, if not then download and install.

## RPATH
Before, when I added the plugin, I had to change the RPATH. Now, I export all the Plugins to the `Plugins` folder. I only need to add `./Plugins` to all the rpaths.
I didn't add it because I don't know how to use it yet, I had to actually do it myself.

Also there seems to be no RPATH on windows, but I've tried it out.
[https://github.com/zmoth/nodeeditor/tree/plugin_extension](https://github.com/zmoth/nodeeditor/tree/plugin_extension)
`/src/PluginsManager.cpp` line:100
```cpp
#if defined(Q_OS_WIN)
#ifdef UNICODE
            SetDllDirectory(folderPath.toStdWString().c_str());
#else
            SetDllDirectory(folderPath.toStdString().c_str());
#endif // !UNICODE
#endif
            loadPluginFromPath(it.filePath());

#if defined(Q_OS_WIN)
            SetDllDirectory(NULL);
#endif
```

>  If you have time, you can try running the [nodeeditor](https://github.com/zmoth/nodeeditor/tree/plugin_extension) with the examples `plugins_load` and `plugins_text`.

## opencv dll copy
ummmm，I restored it. I am using `vcpkg`, which automatically adds relevant DLLs.

## output directory error in Linux
Maybe you need to clean cmake's cache. I had no problem with ubuntu20.04. Both Debug and Release will be placed in `/build/bin/`. And I don't seem to generate the cmake-build-release folder.
The main reason I did this was to have the same output path for Windows, Linux, and Mac, and to be able to separate Debug and Release.

And before that, Windows will produce the following directory: `build/Release`, `build/Debug`
Linux does not differentiate between Release and Debug,They were all put in `build/bin`, and when I compile after Debug, the Release compilation will overwrite the files I Debug.

The output directory is now uniform for all platforms, and all plug-ins will be placed in the Plugins folder.
![image](https://user-images.githubusercontent.com/55644167/221497651-bf323ebb-7341-476b-9328-a934aff6ea5d.png)


So, I still think this is the better way to deal with it.

## console window
I agree, so I've added the `USE_LOG_WINDOW` option, which defaults to `ON`.

---

Look forward to your new suggestions.

zmoth committed:
Modify the opencv package usage logic

zmoth committed:
add USE_LOG_WINDOW option to display the log

rwardlow01 commented:
I think the PR is getting to messy, it might need to be broken up into its separate parts as individual PRs, we have too many issue with various parts and it's going to be difficult to resolve them as a whole.

## USE_LOCAL_OPENCV_PACKAGE
The OpenCV_DIR method is the official way to tell CMake how to find OpenCV:
https://docs.opencv.org/2.4/doc/tutorials/introduction/windows_install/windows_install.html#set-the-opencv-enviroment-variable-and-add-it-to-the-systems-path

in CLion you just add the CMake flag -DOpenCV_DIR=/your/path/to/opencv
and it just works automatically with the CMake build process even if you delete the cache and start over.
this also works great on Linux and Mac using the command line build tools.
with the proposed changes it means I have to set both -DUSE_LOCAL_OPENCV_PACKAGE=ON and -DOpenCV_DIR=/path
if I want to override the auto downloaded version, because I don't use vcpkg and maybe other don't as well.
I'm just trying to make the process simple and use the existing methods already available to set these things and not create redundancies.
there must be something similar for VS now that VS has native CMake support?
If not you should be able to set build environment variables within the IDE specifically for use in the build process without having to create ones at the system level

## output directory error in Linux
The reason I had the build paths setup the way I did before is that it worked correctly across all platforms, with the new process I am getting different results on different platforms and different results depending on if I am building in CLion or with the command line tools.
I don't have any problems with Linux differentiating between release and debug, for each build type they get their own cmake build folder so that must be a VS issue or something?

---

I have to go and look at a few of the other things but with the current issues with these changes I can't easily build without having to do a bunch of manual steps after each build to get files back into the correct place.
For example the resource (app icon) file is still missing from the linux build folder, I had a step in there to copy into the build so the program icon shows up properly when running.
https://github.com/FlowCV-org/FlowCV/blob/893bb6890d75ae0620121df7af5dd97a19c7014b/Editor_UI/CMakeLists.txt#L19
but with the new build folder setup that folder doesn't exist anymore so that also needs to change.
The missing DLLs also needs to be fixed for anyone not using vcpkg or some other work around.

---

I think at this point it's going to be better to tackle these commits as individual PRs.

The OpenCV Download one is almost ready.
The DLL copying can either be put back or add another condition if not using the downloaded OpenCV then it can assume you have the proper paths in place and not copy the DLLs

The build paths and RPath we'll need to discuss some more and figure out a better solution.
The reason Windows doesn't have RPath or need it, is that it has it's own method for loading DLLs and doesn't rely on the RPath setting as you can see in the Plugin code: 
https://github.com/FlowCV-org/FlowCV/blob/893bb6890d75ae0620121df7af5dd97a19c7014b/FlowCV_SDK/third-party/dspatch/src/Plugin.cpp#L100-L105

zmoth commented:
Thanks for the explanation. I will split this pr and I can take my time to perfect them.
